lib/charms/opensearch/v0/helper_cluster.py

```diff
@@ -60,17 +48,79 @@ def suggest_roles(nodes: List[Node], planned_units: int) -> List[str]:
 
-        if nodes_by_roles.get("cluster_manager", 0) >= max_managers:
+        if nodes_by_roles.get("cluster_manager", 0) >= max_cms:
             return base_roles + ["voting_only"]
 
         return base_roles + ["cluster_manager"]
 
+    @staticmethod
+    def node_with_new_roles(remaining_nodes: List[Node]) -> Optional[Node]:
+        """Pick and recompute the roles of the best node to re-balance the cluster."""
+        max_cms, max_voters = ClusterTopology.max_cm_and_voter_nodes(len(remaining_nodes))
+        max_voting_only = max_voters - max_cms
+
+        nodes_by_roles = ClusterTopology.nodes_count_by_role(remaining_nodes)
+        current_cms = nodes_by_roles.get("cluster_manager", 0)
+        current_voting_only = nodes_by_roles.get("voting_only", 0)
+
+        # the nodes involved in the voting are intact, do nothing
+        if current_cms + current_voting_only == max_voters:
+            logger.debug("Suggesting NO changes to the nodes.")
+            return None
+
+        nodes_by_roles = ClusterTopology.nodes_by_role(remaining_nodes)
+
+        if current_cms > max_cms:
+            # remove cm from a node
+            cm = choice(nodes_by_roles["cluster_manager"])
+            logger.debug(f"Suggesting - removal of 'CM': {cm.name}")
+            return Node(cm.name, [r for r in cm.roles if r != "cluster_manager"], cm.ip)
+
+        if current_voting_only > max_voting_only:
+            # remove voting_only from a node
+            voting_only = choice(nodes_by_roles["voting_only"])
+            logger.debug(f"Suggesting - removal of 'voting_only': {voting_only.name}")
+            return Node(
+                voting_only.name,
+                [r for r in voting_only.roles if r != "voting_only"],
+                voting_only.ip,
+            )
+
+        exclude_roles = {"cluster_manager", "voting_only"}
+        data = choice(
+            [node for node in nodes_by_roles["data"] if not exclude_roles & (set(node.roles))]
+        )
+
+        if current_cms < max_cms:
+            # add cm to a data node (that doesn't have voting_only)
+            logger.debug(f"Suggesting - Addition of 'CM' to data: {data.name}")
+            return Node(data.name, data.roles + ["cluster_manager"], data.ip)
+
+        # add voting_only to a data node
+        logger.debug(f"Suggesting - Addition of 'voting_only' to data: {data.name}")
+        return Node(data.name, data.roles + ["voting_only"], data.ip)
+
+    @staticmethod
+    def max_cm_and_voter_nodes(planned_units) -> (int, int):
+        """Get the max number of CMs and voters on a cluster."""
+        max_managers = planned_units
+        max_voters = planned_units
+        if planned_units % 2 == 0:
+            max_managers -= 1
+            max_voters -= 1
+
+        if max_managers > 3:
+            # for a cluster of +3 nodes, we want to have half of the nodes as CMs
+            max_managers = max_managers // 2 + 1
+
+        return max_managers, max_voters
```

Comment:
We need to write specs on manager-eligible roles. I think we agreed on another much simpler scenario the last time because of the new multi-cluster design.

Reply:
Yes - will change this in a future PR - it is tracked in [DPE-1321](https://warthogs.atlassian.net/browse/DPE-1321).